scripts/binary/binary-integrity-check-source.js

```diff
@@ -0,0 +1,168 @@
+const toString = Function.prototype.toString
+const callFn = Function.call
+
+// eslint-disable-next-line no-unused-vars
+const stackIntegrityCheck = function stackIntegrityCheck (options) {
+  const originalStackTraceLimit = OrigError.stackTraceLimit
+  const originalStackTrace = OrigError.prepareStackTrace
+
+  OrigError.stackTraceLimit = Infinity
+
+  OrigError.prepareStackTrace = function (_, stack) {
+    return stack
+  }
+
+  const tempError = new OrigError
+
+  captureStackTrace(tempError, arguments.callee)
+  const stack = tempError.stack.filter((frame) => !frame.getFileName().startsWith('node:internal') && !frame.getFileName().startsWith('node:electron'))
+
+  OrigError.prepareStackTrace = originalStackTrace
+  OrigError.stackTraceLimit = originalStackTraceLimit
+
+  if (stack.length !== options.stackToMatch.length) {
+    throw new Error(`Integrity check failed with expected stack length ${options.stackToMatch.length} but got ${stack.length}`)
+  }
+
+  for (let index = 0; index < options.stackToMatch.length; index++) {
+    const expectedFunctionName = options.stackToMatch[index].functionName
+    const actualFunctionName = stack[index].getFunctionName()
+    const expectedFileName = options.stackToMatch[index].fileName
+    const actualFileName = stack[index].getFileName()
+
+    if (expectedFunctionName && actualFunctionName !== expectedFunctionName) {
+      throw new Error(`Integrity check failed with expected function name ${expectedFunctionName} but got ${actualFunctionName}`)
+    }
+
+    if (expectedFileName && actualFileName !== expectedFileName) {
+      throw new Error(`Integrity check failed with expected file name ${expectedFileName} but got ${actualFileName}`)
+    }
+  }
+}
+
+function validateToString () {
+  if (toString.call !== callFn) {
+    throw new Error('Integrity check failed for toString.call')
+  }
+}
+
+function validateElectron (electron) {
+  if (toString.call(electron.app.getAppPath) !== 'function getAppPath() { [native code] }') {
+    throw new Error(`Integrity check failed for electron.app.getAppPath.toString()`)
+  }
+}
+
+function validateFs (fs) {
+  if (toString.call(fs.readFileSync) !== `function(t,r){const n=splitPath(t);if(!n.isAsar)return g.apply(this,arguments);const{asarPath:i,filePath:a}=n,o=getOrCreateArchive(i);if(!o)throw createError("INVALID_ARCHIVE",{asarPath:i});const c=o.getFileInfo(a);if(!c)throw createError("NOT_FOUND",{asarPath:i,filePath:a});if(0===c.size)return r?"":s.Buffer.alloc(0);if(c.unpacked){const t=o.copyFileOut(a);return e.readFileSync(t,r)}if(r){if("string"==typeof r)r={encoding:r};else if("object"!=typeof r)throw new TypeError("Bad arguments")}else r={encoding:null};const{encoding:f}=r,l=s.Buffer.alloc(c.size),u=o.getFdAndValidateIntegrityLater();if(!(u>=0))throw createError("NOT_FOUND",{asarPath:i,filePath:a});return logASARAccess(i,a,c.offset),e.readSync(u,l,0,c.size,c.offset),validateBufferIntegrity(l,c.integrity),f?l.toString(f):l}`) {
```

Comment:
risk this will ever be different given it is internal node code?

Reply:
This is actually electron code not node code and there's not really a good way for me to get at it at run time. I think we are safe to update this as we update electron (if we need to). I'll add a comment for posterity though.

Reply:
Clarified: [`5c1f981` (#24909)](https://github.com/cypress-io/cypress/pull/24909/commits/5c1f98154e643a23ccddc01e3c02d7700059fcbe)